<img src="../src/packt-banner.png" alt="">

### Exercise 3: Build a SVM modle for Face Recognition Problem
##### (25 points) --> your total will divided by 5 to get 5 points for this exercise.
---

We will use a very famous dataset, called Labelled Faces in the Wild, which
consists of 1288 faces of famous people, and it is available at http://viswww.cs.umass.edu/lfw/lfw-funneled.tgz. 

However, note that it can be easily imported via scikit-learn from the datasets class.
Each image consists of 1850 features: we could proceed by simply using each of them in the model.



Fitting a SVM to non-linear data using the Kernel Trick produces non- linear decision boundaries.
In particular, we seek to:
* Build SVM model with radial basis function (RBF) kernel
* Use a grid search cross-validation to explore ran- dom combinations of parameters.

### Step to do:

1. Loading the dataf from sklearn.datasets:

In [ ]:
from sklearn.datasets import fetch_lfw_people
faces = fetch_lfw_people(min_faces_per_person=60)

2. Since the data can be accessed from the sklearn.datasets module, you need to explore the dataset. 
    - (refer to the first 6 steps in Lab_1 could help you)

a- Print the field names (that is, the keys to the dictionary) (1 point)

In [ ]:
# What fields are in the dictionary?

for key in faces.keys():
    print(key)

b- Print the dataset description contained (2 point)

In [ ]:
# write your code here
# Print the dataset description
description = faces.DESCR
print(description)


3. Print the data, its shape, and the target names. ( 3 points)

In [ ]:
# What does the data look like?
import matplotlib.pyplot as plt

images = faces.images

target_names = faces.target_names

plt.figure(figsize=(10, 5))
for i in range(10): 
    plt.subplot(2, 5, i + 1)
    plt.imshow(images[i], cmap='gray')
    plt.title(target_names[faces.target[i]])
    plt.axis('off')

plt.show()



In [ ]:
# what is the shape of the data?
data = faces.data
data_shape = data.shape
print("Shape of the Data:", data_shape)


In [ ]:
# What is the target names?
target_names = faces.target_names
print("Target Names (Individuals):")
for name in target_names:
    print(name)


4. Divide the data into features (X) using the faces.data and target (y) using faces.target (2 points)

In [ ]:
# Write your code here 
X = faces.data
y = faces.target

5. Splitting the data into training and testing sets. (2 point)

We train the model with 70% of the samples and test with the remaining 30%.

In [ ]:
# Write your code here 

from sklearn.model_selection import train_test_split
X = faces.data
y = faces.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
print("Training set - X:", X_train.shape, "y:", y_train.shape)
print("Testing set - X:", X_test.shape, "y:", y_test.shape)

# print the sizes of our training and test set to verify if the splitting has occurred properly.
print("Training set size (X, y):", X_train.shape, y_train.shape)
print("Testing set size (X, y):", X_test.shape, y_test.shape)


6. Declare SVM model with kernel='rbf', class_weight='balanced' (2 points)

In [ ]:
from sklearn.svm import SVC
from sklearn.svm import SVC
svm_model = SVC(kernel='rbf', class_weight='balanced')
print(svm_model)


7. Use a grid search cross-validationwith 10 CV to explore random combinations of parameters. (3 points) 
    - we will adjust C, which controls the margin
    - and Gamma (γ), which controls the size of the radial basis function kernel, and determine the best model.

In [11]:

from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
from sklearn.datasets import fetch_lfw_people
from sklearn.model_selection import train_test_split
param_grid = {'C': [1, 5, 10, 50], 'gamma': [0.001, 0.0005, 0.01, 0.1]}
#write your code here
faces = fetch_lfw_people(min_faces_per_person=60)

X = faces.data
y = faces.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
svm_model = SVC(kernel='linear', class_weight='balanced')
grid_search = GridSearchCV(estimator=svm_model, param_grid=param_grid, cv=10)
grid_search.fit(X_train, y_train)


print("Best Parameters:", grid_search.best_params_)
print("Best Score:", grid_search.best_score_)


Best Parameters: {'C': 1, 'gamma': 0.001}
Best Score: 0.8358688127379716


8. predict on the test set, using the best model from above step (best_estimator_) (5 points)

In [12]:
# write your code here

best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)
print("Predicted Labels:")
print(y_pred)


Predicted Labels:
[3 4 3 6 6 1 3 3 3 1 3 3 4 3 2 1 3 2 3 2 7 7 5 5 0 3 6 7 3 3 0 6 3 3 3 3 2
 3 3 3 3 3 7 1 3 3 0 1 1 2 7 3 4 6 7 3 7 1 7 0 4 3 7 2 5 4 7 3 4 3 1 3 4 1
 3 4 0 4 3 3 1 3 1 0 7 3 3 2 7 1 1 1 2 3 1 7 7 3 3 1 3 7 1 4 3 3 3 3 7 3 3
 1 0 2 3 3 3 2 4 4 7 7 5 3 3 3 3 3 2 2 3 7 0 3 4 3 4 1 3 1 7 6 5 3 3 1 1 3
 5 4 3 3 7 1 7 1 3 4 1 4 6 1 2 3 2 3 1 7 2 2 1 7 3 3 1 1 1 3 3 1 0 0 1 1 7
 1 1 5 3 4 3 3 4 5 6 3 7 3 3 2 2 3 2 3 3 6 3 3 1 7 3 6 1 2 3 1 1 7 6 3 1 3
 1 7 7 2 7 7 5 7 1 3 3 2 4 4 7 3 3 1 3 4 7 1 4 3 1 1 5 4 2 3 4 1 3 1 2 2 3
 7 3 4 3 7 3 1 3 1 3 2]


9. Model performances:
Run the following code to print the model evaluation metric

In [13]:
from sklearn.metrics import classification_report
labels = list(faces.target_names)
print(classification_report(y_test,y_pred,target_names=labels))

                   precision    recall  f1-score   support

     Ariel Sharon       0.70      0.58      0.64        12
     Colin Powell       0.84      0.84      0.84        51
  Donald Rumsfeld       0.69      0.72      0.71        25
    George W Bush       0.86      0.87      0.86        98
Gerhard Schroeder       0.65      0.81      0.72        21
      Hugo Chavez       1.00      0.67      0.80        15
Junichiro Koizumi       0.91      1.00      0.95        10
       Tony Blair       0.81      0.79      0.80        38

         accuracy                           0.81       270
        macro avg       0.81      0.78      0.79       270
     weighted avg       0.82      0.81      0.81       270



10. What do you observe about the model performances? (5 points)

#### Write your answer here




The model generally has good precision for most classes, with values ranging from 0.65 to 1.00. This suggests that the model's positive predictions are reliable for most classes.
The model also exhibits good recall for most classes, with values ranging from 0.58 to 1.00. This suggests that the model is effective at identifying true positive instances for most classes.
In this case, the F1-scores range from 0.64 to 0.95, which suggests that the model maintains a good balance between precision and recall for most classes.